### DISCUSSION TITLE
Can't rename symbol?

### DISCUSSION THREAD

**Discussion by pepper-lees** [scalameta/nvim-metals#581] [Q&A] [answered]
Hi,

I'm using Metals to add language support for my local instance of [Chipyard](https://github.com/ucb-bar/chipyard). But when I want to rename a symbol (I press `<leader>` -> `l`), the 'Rename symbol' command doesn't appear on the list of available commands. If I open a JS file, then 'Rename current symbol' DOES appear. So it does seem like a language server issue? But I don't know how to diagnose it.

The only options after pressing `<leader>` -> `l` that appear for me are S - Symbols outline and  s - Search symbols.

**Comment by ckipp01** [comment]
Thanks for the questions @JL102 .

> But when I want to rename a symbol (I press <leader> -> l), the 'Rename symbol' command doesn't appear on the list of available commands.

Can you tell me a bit more about this? For rename do you just have `<leader>l` bound to `*vim.lsp.buf.rename()`? What do you mean by available commands?

Also with Chipyard, does the metals doctor, (`:MetalsRunDoctor`) show you any issues or is everything ✅ ?

**Reply by pepper-lees** [reply]
Sorry, I forgot to clarify I'm using AstroNvim, which has a bunch of conveniences pre-configured. When I press the leader key a popup appears on the bottom with the list of keymappings that are available.

Here's an example when I type `<leader>l`, which is the mapping for language features, on a symbol in a JS file:
![image](https://github.com/scalameta/nvim-metals/assets/11487870/afde469e-2d0c-4065-ac9f-8ae7581da199)

And here's what I see if I do the same in a Scala file:
![image](https://github.com/scalameta/nvim-metals/assets/11487870/e0ff7e85-5f99-464d-ae9f-818012241a4a)

As for the doctor, it's a bit unclear: The project is `lbl-async-mem-module`, but it shows a bunch of other modules listed, like `lbl-async-mem-module-test` and `lbl-async-mem-module-build-build`. The first two are all ✅, but `lbl-async-mem-module-build` and `lbl-async-mem-module-build-build` have an ❌for debugging.

The `<leader>lr` keybind seems to be configured by Astro to only appear if the "textDocument/rename" method is supported:

```
  if client.supports_method "textDocument/rename" then
    lsp_mappings.n["<leader>lr"] = {
      function() vim.lsp.buf.rename() end,
      desc = "Rename current symbol",
    }
  end
```

**Reply by pepper-lees** [reply]
If you need more info on what `client` is, it seems like that code is inside a callback for this function:

```
M.on_attach = function(client, bufnr)
```
and it's inside the file `lsp.lua`,

**Reply by ckipp01** [reply]
> As for the doctor, it's a bit unclear: The project is lbl-async-mem-module, but it shows a bunch of other modules listed, like lbl-async-mem-module-test and lbl-async-mem-module-build-build. The first two are all ✅, but lbl-async-mem-module-build and lbl-async-mem-module-build-build have an ❌for debugging.

So just for context, every module will also have a corresponding `-test` module, since that's just an sbt-ism. So that's normal. As for the `-build` and `-build-build`, that's also normal since that is what is saying if you have support in your build and your metabuild, aka your `build.sbt` file and `project/*`. It's also normal to see the ❌ for debugging there, since that isn't supported.

> The <leader>lr keybind seems to be configured by Astro to only appear if the "textDocument/rename" method is supported:

In Metals this is for sure supported, so I have no idea why this wouldn't show up.

Astro is for sure the missing piece for me here, as I'm really unsure how it works. Since nvim-metals is a completely standalone app, it sometimes doesn't work super well with some of these other very opinionated large frameworks. With Astro how are you setting up Metals?

**Reply by pepper-lees** [reply]
Within Astro's `user.lua` file, I have this block in the list of plugins being returned:

```lua
  {
    "scalameta/nvim-metals",
    dependencies = { "nvim-lua/plenary.nvim" },
    event = "User AstroFile",
    config = function()
      local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })
      vim.api.nvim_create_autocmd("FileType", {
        pattern = { "scala", "sbt", "java" },
        callback = function()
          local metals = require "metals"
          metals.initialize_or_attach {}
          local metals_config = metals.bare_config()
          metals_config.init_options.statusBarProvider = "on"
        end,
        group = nvim_metals_group,
      })
    end,
  },
```
In context:

```lua
return {
  {
    "scalameta/nvim-metals",
    dependencies = { "nvim-lua/plenary.nvim" },
    event = "User AstroFile",
    config = function()
      local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })
      vim.api.nvim_create_autocmd("FileType", {
        pattern = { "scala", "sbt", "java" },
        callback = function()
          local metals = require "metals"
          metals.initialize_or_attach {}
          local metals_config = metals.bare_config()
          metals_config.init_options.statusBarProvider = "on"
        end,
        group = nvim_metals_group,
      })
    end,
  },
}
```

Is there a way to try and open the "rename symbol" interface with a Vim command (like `:RenameSymbol` or something)?

**Comment by pepper-lees** [comment, marked as answer]
I got help on the AstroNvim discord server, and it turns out I wasn't attaching it to Astro's language server properly. This is my config section of user.lua now:

```lua
  {
    "scalameta/nvim-metals",
    dependencies = "nvim-lua/plenary.nvim",
    init = function()
      vim.api.nvim_create_autocmd("FileType", {
        pattern = { "scala", "sbt", "java" },
        callback = function() require("metals").initialize_or_attach(require("astronvim.utils.lsp").config "metals") end,
        group = vim.api.nvim_create_augroup("nvim-metals", { clear = true }),
      })
    end,
  },
```

After changing it, now the keybinds work as expected.
